File: modules/iam/src/test/java/com/media/bus/iam/member/repository/MemberRepositoryTest.java
package com.media.bus.iam.member.repository;

import com.media.bus.iam.modules.member.entity.Member;
import com.media.bus.iam.modules.member.entity.enumerated.MemberStatus;
import com.media.bus.iam.modules.member.repository.MemberRepository;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.test.context.ActiveProfiles;

import java.time.OffsetDateTime;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * MemberRepository 테스트 클래스.
 * 엔티티의 PK(UUID) 수동 삽입 및 자동 생성 동작을 검증합니다.
 */
@DataJpaTest
@ActiveProfiles("test")
class MemberRepositoryTest {

    @Autowired
    private JdbcTemplate jdbcTemplate;

    @Autowired
    private MemberRepository memberRepository;

    @Test
    @DisplayName("의도적으로 PK(UUID)를 설정하여 DB에 직접 Insert 할 때, 해당 ID가 잘 반영되어야 한다.")
    void manualInsertTest() {
        // given
        UUID manualId = UUID.randomUUID();
        String sql = """
                        INSERT INTO auth.member (id, login_id, password, email, phone_number, email_verified, status, created_at, updated_at)
                        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?);
                     """;

        // when
        // JPA의 영속성 컨텍스트 관리를 우회하여 저수준(JdbcTemplate)으로 직접 삽입하는 방식입니다.
        // 데이터 마이그레이션이나 레거시 PK 유지보수 시 주로 사용됩니다.
        jdbcTemplate.update(sql,
                manualId,
                "manual_user",
                "password123!",
                "manual@example.com",
                "010-1234-5678",
                false,
                MemberStatus.ACTIVE.name(),
                OffsetDateTime.now(),
                OffsetDateTime.now());

        // then
        // JPA Repository를 통해 조회하여 데이터가 정상적으로 저장되었는지 확인
        Member foundMember = memberRepository.findById(manualId).orElseThrow();
        assertThat(foundMember.getId()).isEqualTo(manualId);
        assertThat(foundMember.getLoginId()).isEqualTo("manual_user");
    }

    @Test
    @DisplayName("JPA를 사용하여 PK를 설정하지 않고 저장할 때, UUID가 자동으로 생성되어야 한다.")
    void saveWithAutomaticIdTest() {
        // given
        Member member = Member.builder()
                .loginId("auto_user")
                .password("password123!")
                .email("auto@example.com")
                .phoneNumber("010-8765-4321")
                .status(MemberStatus.ACTIVE)
                .build();

        // when
        Member savedMember = memberRepository.save(member);

        // then
        assertThat(savedMember.getId()).isNotNull();
        assertThat(savedMember.getLoginId()).isEqualTo("auto_user");
    }
}

File: modules/iam/src/test/kotlin/com/media/bus/iam/member/repository/MemberRepositoryTest.kt
package com.media.bus.iam.member.repository

import com.media.bus.common.entity.common.UuidV7
import com.media.bus.iam.member.entity.MemberEntity
import com.media.bus.iam.member.entity.MemberTable
import com.media.bus.iam.member.entity.enumerated.MemberStatus
import org.assertj.core.api.Assertions.assertThat
import org.jetbrains.exposed.v1.jdbc.Database
import org.jetbrains.exposed.v1.jdbc.SchemaUtils
import org.jetbrains.exposed.v1.jdbc.transactions.transaction
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance

/**
 * ## MemberRepository Exposed 단위 테스트
 *
 * H2 인메모리 DB로 Exposed DAO 동작을 검증한다.
 * - UUID 지정 생성 시 해당 ID가 그대로 저장되는지 확인
 * - UuidV7.generate()로 생성 시 UUID가 자동으로 부여되는지 확인
 */
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class MemberRepositoryTest {

    private lateinit var memberRepository: MemberRepository

    @BeforeAll
    fun setUp() {
        Database.connect(
            url = "jdbc:h2:mem:iamtest;MODE=PostgreSQL;DB_CLOSE_DELAY=-1",
            driver = "org.h2.Driver",
        )
        transaction {
            exec("CREATE SCHEMA IF NOT EXISTS auth")
            SchemaUtils.create(MemberTable)
        }
        memberRepository = MemberRepository()
    }

    @Test
    @DisplayName("UUID를 지정하여 멤버를 생성하면, 해당 ID가 그대로 저장되어야 한다.")
    fun `createWithSpecificUuid_shouldPersistGivenId`() {
        transaction {
            val specificId = UuidV7.generate()
            val member = MemberEntity.new(specificId) {
                loginId = "manual_user"
                password = "password123!"
                email = "manual@example.com"
                phoneNumber = "01012345678"
                emailVerified = false
                status = MemberStatus.ACTIVE
                businessNumber = null
                memberName = "테스트"
            }

            val found = memberRepository.findById(member.id.value)
            assertThat(found).isNotNull
            assertThat(found!!.loginId).isEqualTo("manual_user")
            assertThat(found.id.value).isEqualTo(specificId)
        }
    }

    @Test
    @DisplayName("UuidV7.generate()로 멤버를 생성하면, UUID가 자동으로 부여되어야 한다.")
    fun `createWithGeneratedUuid_shouldHaveNonNullId`() {
        transaction {
            val ts = System.currentTimeMillis()
            val member = MemberEntity.new(UuidV7.generate()) {
                loginId = "auto_user_$ts"
                password = "password123!"
                email = "auto_$ts@example.com"
                phoneNumber = "01087654321"
                emailVerified = false
                status = MemberStatus.ACTIVE
                businessNumber = null
                memberName = "테스트"
            }

            assertThat(member.id.value).isNotNull
            assertThat(member.loginId).startsWith("auto_user_")
        }
    }

    @Test
    @DisplayName("existsByLoginId — 존재하는 loginId는 true를 반환해야 한다.")
    fun `existsByLoginId_existingId_returnsTrue`() {
        transaction {
            val ts = System.currentTimeMillis()
            MemberEntity.new(UuidV7.generate()) {
                loginId = "exists_user_$ts"
                password = "pw"
                email = "exists_$ts@example.com"
                phoneNumber = "01011112222"
                emailVerified = false
                status = MemberStatus.ACTIVE
                businessNumber = null
                memberName = "테스트"
            }

            assertThat(memberRepository.existsByLoginId("exists_user_$ts")).isTrue
            assertThat(memberRepository.existsByLoginId("nonexistent_$ts")).isFalse
        }
    }
}
